galley/pkg/nothingupmysleeve/wait.go

```diff
@@ -0,0 +1,174 @@
+// limitations under the License.
+
+package nothingupmysleeve
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/rest"
+)
+
+func WaitForDistribution(ctx context.Context, namespace string, resourceName string, resourceType string, resourceVersion string,
+	client rest.RESTClient, istioNamespace string, pollInterval time.Duration) error {
+	// TODO: set up watch to add new versions to accepted versions
+	// alternative: though this is not documented, k8s resource versions are almost always increasing ints.
+	// changing the contains() check to a >= check, and that simplifies things.
+	acceptedVersions := []string{resourceVersion}
+	// get list of all pilots
+	pilots, err := findAllPilotInstances(istioNamespace, client)
+	if err != nil {
+		return err
+	}
+	if len(pilots) < 1 {
+		return errors.New("unable to find any Pilot instances")
+	}
+
+	// query each periodically, removing those that succeed.
+	wg := sync.WaitGroup{}
+	wg.Add(len(pilots))
+
+	targetResource := fmt.Sprintf("%s/%s/%s", resourceType, namespace, resourceName)
+	path := fmt.Sprintf("/debug/config_distribution?resource=%s", targetResource)
+	var errs error
+
+	for _, pilot := range pilots {
+		go func() {
+			t := time.NewTicker(pollInterval)
+			for {
+				// do the get here
+				// port is 8080 for 1.6+
+				response, err := proxyGet(pilot.Name, pilot.Namespace, path, 15014, client).DoRaw()
+				if err != nil {
+					// this might not be the right thing to do on error
+					errs = multierror.Append(errs, err)
+					break
+				}
+				var configVersions []SyncedVersions
+				err = json.Unmarshal(response, &configVersions)
+				if err != nil {
+					// this might not be the right thing to do on error
+					errs = multierror.Append(errs, err)
+					break
+				}
+				versionCount := make(map[string]int)
+				for _, configVersion := range configVersions {
+					countVersions(versionCount, configVersion.ClusterVersion)
+					countVersions(versionCount, configVersion.RouteVersion)
+					countVersions(versionCount, configVersion.ListenerVersion)
+				}
+
+				notpresent := 0
+				for version, count := range versionCount {
+					if !contains(acceptedVersions, version) {
+						notpresent += count
+					}
+				}
+				if notpresent < 1 {
+					// success!
+					break
+				} else {
+					fmt.Printf("still waiting for %d proxies from pilot %s", notpresent, pilot.Name)
+				}
+				// repeat every interval until success or cancel
+				select {
+				case <-ctx.Done():
+					errs = multierror.Append(errs, ctx.Err())
+					break
+				case <-t.C:
+					continue
+				}
+			}
+			wg.Done()
+		}()
+	}
+
+	return nil
+}
+
+func findAllPilotInstances(istioNamespace string, client rest.RESTClient) ([]v1.Pod, error) {
+	req := client.Get().
+		Resource("pods").
+		Namespace(istioNamespace)
+
+	req.Param("labelSelector", "istio=pilot")
```

Comment:
Can we start to use `app=istiod` label everywhere. https://github.com/istio/istio/blob/6ba5c6b9a1e7d87e9d153886c363461e73d4df70/pkg/kube/client.go#L470

Reply:
Sorry, this was garbage too.